core/embed/trezorhal/optiga/optiga.c

```diff
@@ -19,8 +19,57 @@
 
 #include "optiga.h"
 #include <string.h>
+#include "hmac.h"
+#include "memzero.h"
 #include "optiga_commands.h"
 
+// PIN counter reset key / Master secret (OID 0xF1D0).
+#define OID_PIN_SECRET (OPTIGA_OID_DATA + 0)
+
+// PIN counter (OID 0xE120).
+#define OID_PIN_COUNTER (OPTIGA_OID_COUNTER + 0)
+
+// Salted PIN (OID 0xF1D4).
+#define OID_SALTED_PIN (OPTIGA_OID_DATA + 4)
+
+// Key for HMAC-SHA256 PIN derivation step (OID 0xF1D1).
+#define OID_PIN_HMAC (OPTIGA_OID_DATA + 1)
+
+// Key for AES-CMAC PIN derivation step (OID 0xE200).
+#define OID_PIN_CMAC OPTIGA_OID_SYM_KEY
+
+// Key for ECDH PIN derivation step (OID 0xE0F3).
+#define OID_PIN_ECDH (OPTIGA_OID_ECC_KEY + 3)
+
+// The number of times optiga_pin_stretch_secret() is called in PIN processing.
+#define PIN_STRETCH_PHASE_COUNT 2
+
+// The number of times the stretching is repeated in each PIN processing phase.
+#define PIN_STRETCH_ITERATIONS 5
```

Comment:
We need to decide on the exact number of iterations. Ideally this would be set so that it takes about one second to complete the PIN verification. The problem is that one PIN verification causes `2 + 6 * PIN_STRETCH_ITERATIONS` security events. That translates to `12 * PIN_STRETCH_ITERATIONS - 6` writes to the security event counter (SEC), assuming the default configuration and that SEC credit is at its maximum when the PIN verification begins. The endurance of the SEC is 1600000 writes. So if we set this constant to `PIN_STRETCH_ITERATIONS=2`, then that allows for a maximum of 88888 PIN verifications over the lifetime of the device. That seems perfectly acceptable. In terms of lifetime, this is equivalent to setting the LUC counter limit [here](https://github.com/trezor/trezor-firmware/pull/3296/files#r1333205454) to `6 * PIN_STRETCH_ITERATIONS * 88888 = 1066656`.

Here are the results for more `PIN_STRETCH_ITERATIONS` values:

`PIN_STRETCH_ITERATIONS` | Max. PIN verifications | Equivalent LUC limit
-- | -- | --
1 | 266,667 | 1,600,000
2 | 88,889 | 1,066,667
3 | 53,333 | 960,000
4 | 38,095 | 914,286
5 | 29,630 | 888,889
6 | 24,242 | 872,727
7 | 20,513 | 861,538
8 | 17,778 | 853,333
9 | 15,686 | 847,059
10 | 14,035 | 842,105

As the table demonstrates, the endurance of the monotonic counter (600,000 writes) is actually not the limiting factor, so we might as well implement the LUC execute access condition for the stretching secrets.

Reply:
Two more things to add:

- Note that there is a natural limit on the number of PIN verifications determined by the `OID_PIN_COUNTER` (600,000 writes). That's at most 300,000 successful unlock operations. Again the limiting factor is the SEC.
- There are situations where we currently unnecessarily involve Optiga in the PIN verification process, e.g. when changing the PIN we first verify and then set the PIN even if the storage is unlocked. I think the only part where it's necessary to involve the Optiga in setting the PIN, because the Optiga adds no security to the PIN verification when the storage is already unlocked. We just have to remember to increment `OID_PIN_COUNTER` if PIN verification fails. This should be handled in a separate issue.